packages/api/src/routes/medical/document.ts

```diff
@@ -108,29 +119,117 @@ router.post(
- * GET /downloadUrl
+/**
+ * Handles the logic for download url endpoints.
  *
+ * @param req Request object.
+ * @returns URL for downloading the document.
+ */
+async function getDownloadUrl(req: Request): Promise<string> {
+  const cxId = getCxIdOrFail(req);
+
+  const fileName = getFromQueryOrFail("fileName", req);
+  const type = getFrom("query").optional("conversionType", req);
+  const conversionType = type ? docConversionTypeSchema.parse(type) : undefined;
+
+  if ((typeof fileName !== "string" || fileName.indexOf(cxId) !== -1) && !Config.isSandbox()) {
+    throw new ForbiddenError();
+  }
+
+  const url = await downloadDocument({ fileName, conversionType });
+  return url;
+}
+
+// TODO: Redirect this endpoint to the new one (download-url)
+/** ---------------------------------------------------------------------------
+ * GET /document/downloadUrl
  * Fetches the document from S3 and sends a presigned URL
+ * @deprecated Use the GET /download-url endpoint instead.
  *
  * @param req.query.fileName The file name of the document in s3.
  * @param req.query.conversionType The doc type to convert to.
  * @return presigned url
  */
 router.get(
   "/downloadUrl",
+  asyncHandler(async (req: Request, res: Response) => {
+    const url = getDownloadUrl(req);
+    return res.status(MOVED_PERMANENTLY).json({ url });
+  })
+);
+
+/** ---------------------------------------------------------------------------
+ * GET /document/download-url
+ *
+ * Fetches the document from S3 and sends a presigned URL
+ *
+ * @param req.query.fileName The file name of the document in s3.
+ * @param req.query.conversionType The doc type to convert to.
+ * @return presigned url
+ */
+router.get(
+  "/download-url",
+  asyncHandler(async (req: Request, res: Response) => {
+    const url = getDownloadUrl(req);
+    return res.status(OK).json({ url });
+  })
+);
+
+/**
+ * POST /document/upload-url
+ *
+ * Uploads a medical document and creates a Document Reference for that file on the FHIR server.
+ *
+ * @param patientId - The ID of the patient.
+ * @body - The Partial<DocumentReference> draft to specify context for the file to be uploaded.
+ *
+ * @return documentReferenceID for the uploaded file.
+ * Refer to Metriport Documentation for more details:
+ * https://docs.metriport.com/medical-api/api-reference/document/post-upload-url
+ */
+router.post(
+  "/upload-url",
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getCxIdOrFail(req);
-    const fileName = getFromQueryOrFail("fileName", req);
-    const fileHasCxId = fileName.includes(cxId);
-    const type = getFrom("query").optional("conversionType", req);
-    const conversionType = type ? docConversionTypeSchema.parse(type) : undefined;
+    const patientId = getFromQueryOrFail("patientId", req);
+    const fileId = uuidv7();
+    const s3FileName = createS3FileName(cxId, patientId, fileId);
+    const docRefId = uuidv7();
+    const s3Key = s3FileName + "_upload_" + docRefId;
```

Comment:
It would be good to have this key being created on the same place it's decomposed later after the file is downloaded. Concentrating related logic in one place helps with maintainability.

Reply:
Agree. But I decided to avoid adding this suffix - we're already going to be using dedicated buckets for uploads